internal/pubsub/subscriber.go

```diff
@@ -2,367 +2,232 @@ package pubsub
+// listen listens for messages on a channel and calls the registered message handler
+func (s Subscriber) listen(messages <-chan *message.Message, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-	_, err := c.js.StreamInfo(c.stream)
-	if err == nil {
-		c.logger.Debugw("stream exists, not recreating", "stream_name", c.stream)
-		return nil
-	}
+	for msg := range messages {
+		if err := s.processEvent(msg); err != nil {
+			s.logger.Warn("Failed to process msg: ", err)
 
-	if !errors.Is(err, nats.ErrStreamNotFound) {
-		return err
+			msg.Nack()
+		} else {
+			msg.Ack()
+		}
 	}
+}
 
-	_, err = c.js.AddStream(&nats.StreamConfig{
-		Name: c.stream,
-		Subjects: []string{
-			c.prefix + ".>",
-		},
-		Storage:   nats.FileStorage,
-		Retention: nats.LimitsPolicy,
-		Discard:   nats.DiscardNew,
-	})
-
-	return err
+// Close closes the subscriber connection and unsubscribes from all subscriptions
+func (s *Subscriber) Close() error {
+	return s.subscriber.Close()
 }
 
-// Listen ensures a stream exists, binds to it, and listens for resource lifecycle events on that stream.
-func (c *Client) Listen() error {
-	if err := c.connect(); err != nil {
-		return err
-	}
+// processEvent event message handler
+func (s *Subscriber) processEvent(msg *message.Message) error {
+	changeMsg, err := events.UnmarshalChangeMessage(msg.Payload)
+	if err != nil {
+		s.logger.Errorw("failed to process data in msg", zap.Error(err))
 
-	// Ensure stream exists before we attempt to listen on it
-	if err := c.ensureStream(); err != nil {
 		return err
 	}
 
-	// Set subscription options. We specifically want manual acks in case something goes wrong
-	// persisting the event
-	subOpts := []nats.SubOpt{
-		nats.BindStream(c.stream),
-		nats.Durable(c.consumer),
-		nats.ManualAck(),
-		nats.AckExplicit(),
-	}
-
-	// For each resource type, we create a single subscription. This means that a single worker
-	// process exists for each resource type, which may be a limiting factor if some resource
-	// subjects are chattier than others
-	for _, name := range c.resourceTypeNames {
-		subject := fmt.Sprintf("%s.%s.>", c.prefix, name)
-		queueName := fmt.Sprintf("%s-%s", c.consumer, name)
+	ctx, span := tracer.Start(context.Background(), "pubsub.receive", trace.WithAttributes(attribute.String("pubsub.subject", changeMsg.SubjectID.String())))
 
-		subscription, err := c.js.QueueSubscribe(subject, queueName, c.receiveMsg, subOpts...)
-		if err != nil {
-			return err
-		}
+	defer span.End()
 
-		c.subscriptions = append(c.subscriptions, subscription)
-	}
+	resource, err := s.qe.NewResourceFromID(changeMsg.SubjectID)
+	if err != nil {
+		s.logger.Warnw("invalid subject id", "error", err.Error())
 
-	return nil
-}
+		msg.Ack()
 
-// Stop drains all subscriptions for the client.
-func (c *Client) Stop() error {
-	for _, sub := range c.subscriptions {
-		err := sub.Drain()
-		if err != nil {
-			return err
-		}
+		return err
```

Comment:
If this branch gets hit then we'll call `msg.Ack()` and return a non-nil error, which will result in `msg.Nack()` being called above. This should probably just log and return nil.